# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю

Выполнили студенты гр. 2381 Долотов Никита и Зазуля Илья. Вариант №18

## Цель работы

Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».

## Основные теоретические положения

Корреляционный момент определяется как:
$$\mu_{xy}=M\{[x - M(X)] \cdot [y - M(Y)]\}$$

Коэффициент корреляции определяется как:
$$r_{xy}=\frac{\mu_{xy}}{\sigma_x\sigma_y}$$

Для коэффициента корреляции справедливо соотношение:
$$|r_{xy}| \leq 1$$

Случайные величины $X$ и $Y$ называют коррелированными, если их корреляционный момент или их коэффициент корреляции отличен от нуля. В противном случае эти величины некоррелированы. \
Если случайные величины $X$ и $Y$ коррелированы, то они зависимы.

Значение $\bar r_{xy}$ - статистической оценки $r_{xy}$ - коэффициента корреляции можно вычислить по формуле:
$$\bar r_{xy} = \frac{\sum\limits_{i=1}^{K_y}\sum\limits_{i=1}^{K_x} n_{ij} y_i x_j}{NS_xS_y}$$

При вычислении выборочного коэффициента корреляции удобно использовать условные варианты:
$$\bar r_{xy} = \frac{\sum\limits_{i=1}^{K_y}\sum\limits_{i=1}^{K_x} n_{ij} u_i v_j - N \bar u_{\text в} \bar v_{\text в} }{N \bar\sigma_u \bar\sigma_v}$$

С помощью преобразования Фишера можно перейти к случайно величине z:
$$\bar z = 0.5 \cdot ln \frac{1 + \bar r_{xy}}{1 - \bar r_{xy}}, (4.1)$$

Распределение $z$ при неограниченном возрастании объема выборки асимпототически нормальное со значением СКВО, равным:
$$\bar\sigma_z = \frac{1}{\sqrt{N-3}}, (4.2)$$

Доверительный интервал генеральное совокупности с доверительной вероятностью $\gamma$ определяют по следующей схеме:
1. По формуле (4.1) вычисляют выборочное значение $\bar z$;
2. По формуле (4.2) вычисляют значение $\bar\sigma_z$;
3. Доверительный интервал для генерального значения представляется в виде:
$$(\bar z - \lambda (\gamma) \bar\sigma_z, \bar z + \lambda (\gamma) \bar\sigma_z),$$
где значение $\lambda (\gamma)$ должно удовлетворять условию:
$$\Phi[\lambda (\gamma)] = \frac{\gamma}{2}$$
4. Для пересчета интервала в доверительный интервал для коэффициента корреляции с тем же значением $\gamma$ необходимо воспользоваться обратным преобразованием Фишера:
$$r = th(z)$$

Так как $\bar r_{xy}$ является случайной величиной, то это еще не значит что $r_{xy}$ - коэффициент корреляции для генеральной совокупности тоже отличен от нуля. \
Возникает необходимость проверить гипотезу $H_0: r_{xy} = 0$. \
Альтернативной будет гипотеза $H_1: r_{xy} \neq 0$. \
Если основная гипотеза $H_0$ отвергается, то это означает, что выборочный коэффициент корреляции $\bar r_{xy}$ значимо отличается от нуля (значим). В противном случае - $\bar r_{xy}$ незначим.

В качестве критерия проверки статистической гипотезы о значимости выборочного коэффициента корреляции можно принять случайную величину:
$$T = \frac{\bar r_{xy} \sqrt{N-2}}{\sqrt{1 - \bar r_{xy}^2}}$$


## Постановка задачи

Из заданной генеральной совокупности сформировать выборку по второму признаку. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса и коэффициента вариации). Для системы двух случайных величин X (первый признак) и Y (второй признак) сформировать двумерную выборку и найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

In [2]:
import pandas as pd
import numpy as np
from IPython.display import display, Math, HTML, Markdown

Отобразим таблицу по второму признаку из первой лабораторной работы

In [32]:
N = 104
bins = np.array([86.80, 100.41, 114.02, 127.64, 141.25, 154.86, 168.47, 182.09, 195.70])
x_mids = np.array([93.61, 107.22, 120.83, 134.44, 148.06, 161.67, 175.28, 188.89])
m = np.array([9, 14, 25, 25, 13, 10, 7, 1])
m_rel = np.array([0.0865, 0.1346, 0.2404, 0.2404, 0.1250, 0.0962, 0.0673, 0.0096])
m_accum = np.array([9, 23, 48, 73, 86, 96, 103, 104])

rows = []
for i in range(len(bins) - 1):
    left, right = bins[i], bins[i + 1]
    rows.append({
        "i": i + 1,
        "[x_i, x_{i+1})": f"[{left:.2f}, {right:.2f})",
        "x̃_i": round((left + right) / 2, 2),
        "m_i": m[i],
        "~m_i": round(m[i] / N, 4),
    })

int_df = pd.DataFrame(rows)
int_df.loc[len(int_df)] = {
    "i": "Σ", "[x_i, x_{i+1})": "", "x̃_i": "",
    "m_i": int_df["m_i"].sum(),
    "~m_i": round(int_df["~m_i"].sum(), 4)
}

print("\nИнтервальный ряд:")
display(int_df.set_index("i"))


Интервальный ряд:


,"[x_i, x_{i+1})",x̃_i,m_i,~m_i
i,,,,
1,"[86.80, 100.41)",93.6,9,0.0865
2,"[100.41, 114.02)",107.22,14,0.1346
3,"[114.02, 127.64)",120.83,25,0.2404
4,"[127.64, 141.25)",134.44,25,0.2404
5,"[141.25, 154.86)",148.06,13,0.1250
6,"[154.86, 168.47)",161.67,10,0.0962
7,"[168.47, 182.09)",175.28,7,0.0673
8,"[182.09, 195.70)",188.9,1,0.0096
Σ,,,104,1.0000


### Задание 1

In [34]:
C = 134.44
h = 13.61

n = m.copy()
u = np.array([(x_mid - C) / h for x_mid in x_mids], dtype=float)

rows = []

for i in range(len(u)):
    rows.append({
        "x_i": x_mids[i],
        "n_i": n[i],
        "u_i": round(u[i]),
        "n_i*u_i": n[i] * round(u[i]),
        "n_i*u_i^2": n[i] * round(u[i])**2,
        "n_i*u_i^3": n[i] * round(u[i])**3,
        "n_i*u_i^4": n[i] * round(u[i])**4,
        "n_i*(u_i+1)^4": n[i] * (round(u[i]) + 1)**4,
    })

df = pd.DataFrame(rows)
df.index = range(1, len(df) + 1)
df.loc["Σ"] = {
    "x_i": "-",
    "n_i": df["n_i"].sum(),
    "u_i": "-",
    "n_i*u_i": df["n_i*u_i"].sum(),
    "n_i*u_i^2": df["n_i*u_i^2"].sum(),
    "n_i*u_i^3": df["n_i*u_i^3"].sum(),
    "n_i*u_i^4": df["n_i*u_i^4"].sum(),
    "n_i*(u_i+1)^4": df["n_i*(u_i+1)^4"].sum(),
}

df

,x_i,n_i,u_i,n_i*u_i,n_i*u_i^2,n_i*u_i^3,n_i*u_i^4,n_i*(u_i+1)^4
1,93.61,9,-3,-27,81,-243,729,144
2,107.22,14,-2,-28,56,-112,224,14
3,120.83,25,-1,-25,25,-25,25,0
4,134.44,25,0,0,0,0,0,25
5,148.06,13,1,13,13,13,13,208
6,161.67,10,2,20,40,80,160,810
7,175.28,7,3,21,63,189,567,1792
8,188.89,1,4,4,16,64,256,625
Σ,-,104,-,-22,294,-34,1974,3618


In [5]:
M = [1 / N * (n * (u ** i)).sum() for i in range(1, 5)]

m = [
    0,
    ( M[1] - M[0]**2 ) * h**2,
    ( M[2] - 3*M[1]*M[0] + 2*(M[0]**3) ) * h**3,
    ( M[3] - 4*M[2]*M[0] + 6*M[1]*(M[0]**2) - 3*(M[0]**4) ) * h**4,
]

rows = []

for i in range(len(M)):
  rows.append({
      "M": M[i],
      "m": m[i],
  })

df = pd.DataFrame(rows)
df.index = range(1, len(df) + 1)
df

,M,m
1,-0.211319,0.000000
2,2.827743,515.517024
3,-0.324124,3654.639381
4,18.990551,667973.646525


In [6]:
xv = 1 / N * np.sum(n * x_mids)
D = np.sum(n * (x_mids - xv)**2) / N
sigma_v = np.sqrt(D)
s_2 = N / (N - 1) * D
s = np.sqrt(s_2)

A_s = m[2] / s**3
E = m[3] / s**4 - 3

i = 3
left = 114.02
right = 127.64

moda = left + h * (m_rel[i] - m_rel[i-1]) / ( (m_rel[i] - m_rel[i-1]) + (m_rel[i] - m_rel[i+1]) )
median = left + h / m_rel[i] * (0.5 - np.sum([m_rel[j] for j in range(i)]))
V = sigma_v / np.abs(xv) * 100


results = pd.DataFrame([
    {"Показатель": "Выборочное среднее", "Обозначение": r"$\bar x$", "Значение": xv},
    {"Показатель": "Дисперсия", "Обозначение": "$D$", "Значение": D},
    {"Показатель": "Среднеквадратичное отклонение", "Обозначение": r"$\sigma_{в}$", "Значение": sigma_v},
    {"Показатель": "Исправленная дисперсия", "Обозначение": "$S^2$", "Значение": s_2},
    {"Показатель": "Исправленное СКО", "Обозначение": "$S$", "Значение": s},
    {"Показатель": "Асимметрия", "Обозначение": "$A_s$", "Значение": A_s},
    {"Показатель": "Эксцесс", "Обозначение": "$E$", "Значение": E},
    {"Показатель": "Мода", "Обозначение": "$M_o^*$", "Значение": moda},
    {"Показатель": "Медиана", "Обозначение": "$M_e^*$", "Значение": median},
    {"Показатель": "Коэффициент вариации, %", "Обозначение": "$V$", "Значение": V},
])

results["Значение"] = results["Значение"].round(4)

display(Markdown(results.to_markdown(index=False)))

| Показатель                    | Обозначение   |   Значение |
|:------------------------------|:--------------|-----------:|
| Выборочное среднее            | $\bar x$      |   131.564  |
| Дисперсия                     | $D$           |   515.517  |
| Среднеквадратичное отклонение | $\sigma_{в}$  |    22.705  |
| Исправленная дисперсия        | $S^2$         |   520.522  |
| Исправленное СКО              | $S$           |    22.815  |
| Асимметрия                    | $A_s$         |     0.3077 |
| Эксцесс                       | $E$           |    -0.5346 |
| Мода                          | $M_o^*$       |   114.02   |
| Медиана                       | $M_e^*$       |   116.2    |
| Коэффициент вариации, %       | $V$           |    17.2578 |

По результатам расчётов выборочное среднее равно $131.56$, а среднеквадратичное отклонение — $22.71$, что говорит об умеренном разбросе значений относительно среднего. Коэффициент вариации $V=17.26%$ меньше $33\%$, следовательно, совокупность можно считать достаточно однородной. Асимметрия положительна
($A_s=0.31$), значит распределение слабо смещено вправо. Эксцесс отрицателен
($E=−0.53$), следовательно, распределение менее островершинное, чем нормальное.

### Задание 2

Интервалы были сформированы в первой лабораторной работе

In [8]:
X = np.array([86.80, 100.41, 114.02, 127.64, 141.25, 154.86, 168.47, 182.09, 195.70])
Y = np.array([344.00, 378.88, 413.75, 448.62, 483.50, 518.38, 553.25, 588.12, 623.00])

x_mids = np.round((X[:-1] + X[1:]) / 2, 2)
y_mids = np.round((Y[:-1] + Y[1:]) / 2, 2)

x_labels = [round(X[i], 2) for i in range(len(X)-1)]
y_labels = [round(Y[i], 2) for i in range(len(Y)-1)]

df = pd.read_csv("sample_104.csv")

df["E"] = pd.cut(df["E"], bins=X, labels=x_labels, right=True, include_lowest=True)
df["nu"] = pd.cut(df["nu"], bins=Y, labels=y_labels, right=True, include_lowest=True)

# строим двумерный интервальный вариационный ряд
table = pd.crosstab(df["nu"], df["E"])

# добавляем частоты
table["n_y"] = table.sum(axis=1)
table.loc["n_x"] = table.sum(axis=0)

display(table)

E,86.8,100.41,114.02,127.64,141.25,154.86,168.47,182.09,n_y
nu,,,,,,,,,
344.0,3,0,0,0,0,0,0,0,3
378.88,4,7,6,0,0,0,0,0,17
413.75,2,7,15,5,0,0,0,0,29
448.62,0,0,4,13,5,0,0,0,22
483.5,0,0,0,7,6,5,2,0,20
518.38,0,0,0,0,2,4,2,0,8
553.25,0,0,0,0,0,0,3,0,3
588.12,0,0,0,0,0,1,0,1,2
n_x,9,14,25,25,13,10,7,1,104


### Задание 3

In [13]:
corr_table = pd.DataFrame(
    freq.astype(int),
    index=np.round(y, 2),
    columns=np.round(x, 2)
)

# правые столбцы
X_i = (freq * x).sum(axis=1)
y_i_X_i = y * X_i

corr_table["Xi"] = np.round(X_i, 2)
corr_table["yiXi"] = np.round(y_i_X_i, 2)

# нижние строки
Y_j = (freq * y[:, None]).sum(axis=0)
x_j_Y_j = x * Y_j

bottom1 = pd.Series(
    list(np.round(Y_j, 2)) + ["", np.round(x_j_Y_j.sum(), 2)],
    index=corr_table.columns,
    name="Yj"
)

bottom2 = pd.Series(
    list(np.round(x_j_Y_j, 2)) + [np.round(x_j_Y_j.sum(), 2), ""],
    index=corr_table.columns,
    name="xjYj"
)

corr_table = pd.concat([corr_table, bottom1.to_frame().T, bottom2.to_frame().T])

display(corr_table)

,93.6,107.22,120.83,134.44,148.06,161.67,175.28,188.9,Xi,yiXi
361.44,3,0,0,0,0,0,0,0,280.81,101497.77
396.32,4,7,6,0,0,0,0,0,1849.9,733145.1
431.18,2,7,15,5,0,0,0,0,3422.39,1475683.23
466.06,0,0,4,13,5,0,0,0,2971.38,1384841.36
500.94,0,0,0,7,6,5,2,0,2988.33,1496974.03
535.82,0,0,0,0,2,4,2,0,1293.33,692985.61
570.68,0,0,0,0,0,0,3,0,525.84,300089.0
605.56,0,0,0,0,0,1,0,1,350.56,212285.11
Yj,3531.95,5792.5,10709.9,11721.28,6407.57,5253.52,3785.56,605.56,,6397501.23
xjYj,330608.18,621042.89,1294077.82,1575868.16,948672.78,849310.31,663533.83,114387.26,6397501.23,


По этой корреляционной таблице видно, что между признаками $E$ и $nu$ есть выраженная прямая связь: при увеличении $E$ значения $nu$ тоже в целом увеличиваются. Основная масса частот сосредоточена вдоль диагонали от меньших значений к большим, а в противоположных углах таблицы частоты почти отсутствуют. Это говорит о достаточно тесной положительной зависимости между признаками.

Также видно, что наибольшая концентрация наблюдений приходится на центральные интервалы, значит распределение пар наблюдений наиболее плотное в средней области значений.

In [26]:
freq = table.iloc[:-1, :-1].to_numpy(dtype=float)

x = (X[:-1] + X[1:]) / 2   # середины интервалов E
y = (Y[:-1] + Y[1:]) / 2   # середины интервалов nu

n = freq.sum()

# частоты по X и Y
n_x = freq.sum(axis=0)
n_y = freq.sum(axis=1)

# выборочные средние
x_bar = (n_x * x).sum() / n
y_bar = (n_y * y).sum() / n

# вторые моменты
x2_bar = (n_x * x**2).sum() / n
y2_bar = (n_y * y**2).sum() / n

# смешанный момент
xy_bar = (freq * np.outer(y, x)).sum() / n

# дисперсии и ковариация
D_x = x2_bar - x_bar**2
D_y = y2_bar - y_bar**2
cov_xy = xy_bar - x_bar * y_bar

r_standard = cov_xy / np.sqrt(D_x * D_y)

print("Способ 1: стандартная формула")
display(Math(rf"\bar x = {round(x_bar, 6)}"))
display(Math(rf"\bar y = {round(y_bar, 6)}"))
display(Math(rf"D_x = {round(D_x, 6)}"))
display(Math(rf"D_y = {round(D_y, 6)}"))
display(Math(rf"cov_{{xy}} = {round(cov_xy, 6)}"))
display(Math(rf"r = {round(r_standard, 6)}"))


# Коэффициент корреляции через условные варианты

h_x = x[1] - x[0]
h_y = y[1] - y[0]

# условные начала отсчёта
A_x = x[len(x)//2]
A_y = y[len(y)//2]

u = (x - A_x) / h_x
v = (y - A_y) / h_y

u_bar = (n_x * u).sum() / n
v_bar = (n_y * v).sum() / n

u2_bar = (n_x * u**2).sum() / n
v2_bar = (n_y * v**2).sum() / n

uv_bar = (freq * np.outer(v, u)).sum() / n

r_cond = (uv_bar - u_bar * v_bar) / np.sqrt((u2_bar - u_bar**2) * (v2_bar - v_bar**2))

print("\nСпособ 2: условные варианты")
display(Math(rf"A_x = {A_x} \quad h_x = {round(h_x, 6)}"))
display(Math(rf"A_y = {A_y} \quad h_y = {round(h_y, 6)}"))
display(Math(rf"\bar u = {round(u_bar, 6)}"))
display(Math(rf"\bar v = {round(v_bar, 6)}"))
display(Math(rf"r = {round(r_cond, 6)}"))

print("\nПроверка:")
print("Разница |r1 - r2| =", abs(r_standard - r_cond))

Способ 1: стандартная формула


<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>


Способ 2: условные варианты


<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>


Проверка:
Разница |r1 - r2| = 5.662137425588298e-15


Результаты, найденные двумя способами, совпадают, следовательно, расчёты выполнены верно. Значение коэффициента корреляции $r \approx 0.877$ свидетельствует о сильной положительной линейной связи между признаками
$E$ и $nu$. Это означает, что при увеличении одного признака другой также имеет выраженную тенденцию к увеличению.

In [28]:
from scipy.stats import norm, t

r = r_standard

gammas = [0.95, 0.99]
rows = []

z = np.arctanh(r)
se = 1 / np.sqrt(N - 3)

for gamma in gammas:
    u = norm.ppf((1 + gamma) / 2)

    z_low = z - u * se
    z_high = z + u * se

    rho_low = np.tanh(z_low)
    rho_high = np.tanh(z_high)

    rows.append({
        "γ": gamma,
        "u": round(u, 4),
        "Нижняя граница": rho_low,
        "Верхняя граница": rho_high
    })

ci_df = pd.DataFrame(rows)
ci_df["Нижняя граница"] = ci_df["Нижняя граница"].round(4)
ci_df["Верхняя граница"] = ci_df["Верхняя граница"].round(4)

display(ci_df)

,γ,u,Нижняя граница,Верхняя граница
0,0.95,1.9600,0.8234,0.9150
1,0.99,2.5758,0.8027,0.9244


В обоих случаях доверительные интервалы полностью лежат в области положительных значений и не содержат ноль. Это подтверждает наличие устойчивой положительной связи между признаками. При увеличении доверительной вероятности интервал становится шире, что соответствует теории.

In [33]:
alpha = 0.05

t_obs = r * np.sqrt(n - 2) / np.sqrt(1 - r**2)
t_crit = t.ppf(1 - alpha/2, df=n-2)

display(Math(rf"r = {round(r, 6)}"))
display(Math(rf"t_{{obs}} = {round(t_obs, 6)}"))
display(Math(rf"t_{{crit}} = {round(t_crit, 6)}"))

if abs(t_obs) > t_crit:
    print("Гипотеза H0: r = 0 отвергается.")
    print("Коэффициент корреляции статистически значим.")
else:
    print("Нет оснований отвергнуть гипотезу H0")
    print("Статистически значимая корреляция не подтверждена.")

<IPython.core.display.Math object>

<IPython.core.display.Math object>

<IPython.core.display.Math object>

Гипотеза H0: r = 0 отвергается.
Коэффициент корреляции статистически значим.


Нулевая гипотеза о равенстве коэффициента корреляции нулю отвергается. Следовательно, коэффициент корреляции статистически значим и линейная связь между признаками $E$ и $nu$ действительно существует. Причём, с учётом ранее полученного значения $r \approx 0.877$, эта связь является сильной и положительной.

## Выводы

В ходе выполнения практической работы была проведена статистическая обработка второй выборки и получены точечные оценки основных параметров распределения исследуемого признака. \
Далее по двум признакам был построен двумерный интервальный вариационный ряд и на его основе составлена корреляционная таблица. Анализ таблицы показал, что основные частоты сосредоточены вдоль диагонального направления, что указывает на наличие выраженной прямой зависимости между признаками: при увеличении одного признака второй также имеет тенденцию к увеличению. Это подтверждается значением выборочного коэффициента корреляции, вычисленного двумя способами — по стандартной формуле и методом условных вариант. В обоих случаях получено одно и то же значение: $r \approx 0.877$.
Следовательно, между исследуемыми признаками существует сильная положительная линейная связь. \
Для генерального коэффициента корреляции были построены доверительные интервалы при уровнях доверия $\gamma = 0.95$ и $\gamma = 0.99$. \
Оба интервала полностью лежат в области положительных значений и не содержат ноль, что подтверждает устойчивый положительный характер связи между признаками. При увеличении доверительной вероятности интервал расширяется, что соответствует теоретическим свойствам интервальных оценок. \
Проверка статистической гипотезы $H_0$ при уровне значимости $\alpha = 0.05$ показала, что наблюдаемое значение статистики $t_{obs} \approx 18.43$ значительно превышает критическое значение $t_{crit} \approx 1.983$. Следовательно, нулевая гипотеза отвергается, и коэффициент корреляции статистически значимо отличается от нуля.